gkctl/main.c

```diff
@@ -0,0 +1,259 @@
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+static int
+load_file_to_buffer(const char *file_name, char *buffer, int n)
+{
+	int ret = 0;
+	FILE *fp;
+
+	if ((fp = fopen(file_name, "r")) == NULL) {
+		fprintf(stderr, "Error: Failed to open file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	/*
+	 * Return value equals the number of bytes transferred
+	 * only when size (i.e., second parameter) is 1.
+	 */
+	ret = fread(buffer, 1, n, fp);
+	if (ret == 0 && ferror(fp)) {
+		fprintf(stderr, "Error: %s\n", strerror(errno));
+		ret = -1;
+	} else if (!feof(fp)) {
+		fprintf(stderr, "Error: Failed to read the whole file %s\n",
+			file_name);
+		ret = -1;
+	}
+
+	fclose(fp);
+out:
+	return ret;
+}
+
+static int
+write_all(int conn_fd, const char *msg_buff, int nbytes)
+{
+	int send_size;
+	int tot_size = 0;
+
+	if (nbytes == 0)
+		return 0;
+
+	while ((send_size = write(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += send_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the server is closed. */
+	if (send_size == 0) {
+		fprintf(stderr, "Server disconnected\n");
+		return -1;
+	}
+
+	if (send_size < 0) {
+		fprintf(stderr, "Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+read_all(int conn_fd, char *msg_buff, int nbytes)
+{
+	int recv_size;
+	int tot_size = 0;
+
+	while ((recv_size = read(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += recv_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the server is closed. */
+	if (recv_size == 0) {
+		fprintf(stderr, "Server disconnected\n");
+		return -1;
+	}
+
+	if (recv_size < 0) {
+		fprintf(stderr, "Failed to read data from the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	char send_buff[MSG_MAX_LEN + 2];
+	char recv_buff[MSG_MAX_LEN];
+	size_t len = 0;
+	size_t total_file_len = 0;
+	struct sockaddr_un serv_addr; 
+
+    	struct args args = {
+		/* Defaults. */
+		.server_path = "/tmp/dyn_cfg.socket",
+	};
+
+	/* Read parameters. */
+	argp_parse(&argp, argc, argv, 0, NULL, &args);
+
+	if (sizeof(serv_addr.sun_path) <= strlen(args.server_path)) {
+		fprintf(stderr, "Error: passing a too long server path (i.e., > %lu) - %s!\n",
+			sizeof(serv_addr.sun_path), args.server_path);
+		ret = -1;
+		goto out;
+	}
+	serv_addr.sun_family = AF_UNIX;
+	strcpy(serv_addr.sun_path, args.server_path);
+
+	ret = load_file_to_buffer(args.lua_script_path,
+		send_buff + sizeof(uint16_t), MSG_MAX_LEN);
+	if (ret <= 0) {
```

Comment:
While weird, wouldn't `ret == 0` be valid? If not, we should warn that the file is empty.